### PR TITLE
Vault Configuration Client

### PR DESCRIPTION
**Pull request by thiagolocatelli**
This Pull Request addresses the issue #904 created by @alvarosanchez 

It contains the integration with Vault to load secrets from Vault as application properties. Vault is a little different from Spring Cloud Config, which returns properties for all property sources matched with the application name and environment names in one response. In Vault, for each environment name a request needs to be made. Also, the user can store secrets under the application name only, without any `profile`. For example, if the application name is `vault-config-sample` and the Micronaut environments are `dev` and `docker` and the user created the secrets in a KV store named `backendv1`, the following requests will  be made:

- http://localhost:8200/v1/backendv1/application
- http://localhost:8200/v1/backendv1/vault-config-sample
- http://localhost:8200/v1/backendv1/application/docker
- http://localhost:8200/v1/backendv1/vault-config-sample/docker
- http://localhost:8200/v1/backendv1/application/dev/
- http://localhost:8200/v1/backendv1/vault-config-sample/dev

At least one request is always made, using the application name. The vault client also respects the order, so in this example, dev takes precedence over docker which takes precedence over the application name. If one environment name, aka profile, does not exist in Vault, the error message is displayed in the logs and the next profiles are evaluated (considering `vault.client.fail-fast` is set to `false`). The logs might show a different order of profiles being retrieved from Vault, that's due to the `Flowable.flatMap()` that loses the order when processing items.

I have also updated the documentation and included the entries for this feature.

### CODE REVIEW

**Comment by thiagolocatelli** [comment]
@jameskleeh Thank you for the time you spent helping me with code refactoring and guidance. Much appreciated.

File: src/main/docs/guide/cloud/cloudConfiguration/distributedConfigurationVault.adoc [hunk 1]
```diff
@@ -0,0 +1,47 @@
+Since 1.1, Micronaut features a native https://www.vaultproject.io/[HashiCorp Vault] for those who have not switched to a dedicated more complete solution like Consul.
```
Comment (graemerocher) [inline src/main/docs/guide/cloud/cloudConfiguration/distributedConfigurationVault.adoc]: Since 1.2 I think

Comment (thiagolocatelli) [reply src/main/docs/guide/cloud/cloudConfiguration/distributedConfigurationVault.adoc]: Fixed

**Comment by graemerocher** [comment]
Thanks again for the contribution! Great work!

**Pull request merged**